### ISSUE TITLE
Does sub_feature seed have something to do with feature names?

### ISSUE THREAD

**Issue by PayneJoe** [Microsoft/LightGBM#720]
I am solving regression problems with python interface.

case A:
I set sub_feature 0.8 with default feature_fraction_seed 2, my data set columns is like ['age', 'gender', 'income', 'location', etc...], and get a model M1 after training.

case B:
I set sub_feature 0.8 with default feature_fraction_seed 2, I relocated(rename) one of the columns as below

_data['age1'] = data['age']
data.drop('age', axis= 1, inplace= True)
the columns will be like ['gender', 'income', 'location', etc..., 'age1']_

and get a model M2 after training.

The performance of model M1 and M2 differentiate much. The whole data does not change except its column order(column name changed), why does it happen? 

In case A, I changed feature_fraction_seed into 3,5,10,2017 respectively, and the performance changed much. So, the feature_fraction_seed has something to do with feature names? But the feature_fraction_seed should not have impact on my performance, right? And I cannot figure it out why https://github.com/Microsoft/LightGBM/issues/116 

I tried the "sample features" of another boosting tree algorithm(gbr), the result happened again. 

Any tips or references provided will much appreciated, thanks!

**Comment by guolinke** [comment]
@PayneJoe 
can you provide a example code for case A and B ?

**Comment by PayneJoe** [comment]
Thanks for your so early response, @guolinke 

My productive code is little bit complicated, I will show you the key parts as below:

for case A: 

` _params = {
        'max_bin': 8,
        'boosting_type': 'gbdt',
        'objective': 'regression',
        'metric': 'mae',
        'sub_feature': 0.80,
        #'feature_fraction_seed': 2,
        'bagging_fraction':  0.80,
        #'bagging_seed': 2017,
        'num_leaves': 128,
        'min_data':  200,
        'min_hessian':  0.01,
        'learning_rate': 0.02,
        'bagging_freq': 20
    }
...
 X = TrainData.drop('y', axis= 1)
 Y = TrainData['y']

 d_cv = lightgbm.Dataset(X,label=Y)    
... 
  self._model = lightgbm.train(self._params,
                                     d_cv,
                                     self._iter,
                                     verbose_eval= True)`

My local MAE is 0.064435

For case B:

` _params = {
        'max_bin': 8,
        'boosting_type': 'gbdt',
        'objective': 'regression',
        'metric': 'mae',
        'sub_feature': 0.80,
        #'feature_fraction_seed': 2,
        'bagging_fraction':  0.80,
        #'bagging_seed': 2017,
        'num_leaves': 128,
        'min_data':  200,
        'min_hessian':  0.01,
        'learning_rate': 0.02,
        'bagging_freq': 20
    }
...
 TrainData['tmp'] = TrainData['latitude']  ## rename latitude column, the new column tmp will be appended 
 TrainData.drop('latitude', axis= 1, inplace= True) ## drop the initial column latitude

 X = TrainData.drop('y', axis= 1)
 Y = TrainData['y']

 d_cv = lightgbm.Dataset(X,label=Y)   
... 
  self._model = lightgbm.train(self._params,
                                     d_cv,
                                     self._iter,
                                     verbose_eval= True)`

My local MAE is 0.064450

Thanks again!

**Comment by guolinke** [comment]
I think this difference is small. 

using sub_feature will randomly sample features, which will be affected by the order of features. 
But it has nothing to do with feature names.

**Comment by PayneJoe** [comment]
Hi, @guolinke, but how can I get rid of it because it's really a big difference in my condition.

And, you just mentioned ORDER, you mean it will be okay if I put the new column 'tmp'  in the original position?

**Comment by guolinke** [comment]
yes, it is okay. 

And you can ensemble multi gbdt with different feature_fraction_seed to reduce the variance.

**Comment by PayneJoe** [comment]
Another question, why performance should changes while adjusting parameter feature_fraction_seed? As https://github.com/Microsoft/LightGBM/issues/116 put.

**Comment by PayneJoe** [comment]
Ooh, 'ensemble multi gbdt with different feature_fraction_seed' is great to hear, I will try it later. Thanks @guolinke

**Comment by guolinke** [comment]
it should change if changing feature_fraction_seed. 
different feature_fraction_seed will generate different sample features.

**Comment by PayneJoe** [comment]
The vibration will increase if I enlarge the feature space with the same feature_fraction_seed?

**Comment by guolinke** [comment]
yeah, it could be. But running it with enough iteration and early stopping also can reduce its variance.

**Comment by PayneJoe** [comment]
It seems that parameter tuning would be an extremely tough job with a large feature space.

**Comment by Laurae2** [comment]
@PayneJoe it is perfectly normal to have a performance difference when re-ordering features / rows / changing seed. In addition, in LightGBM, it is also dependent on the number of threads used.

**Comment by PayneJoe** [comment]
I think I underestimated these stuffs, and I will take it easy when using boosting/bagging tree algorithms. Thanks @Laurae2 and @guolinke , you can close this issue later on.
